File: backend/src/modules/gateway/upstream/header-policy.ts
/**
 * Header policy is deliberately separate from session resolution.
 *
 * A header can be request-dynamic or sensitive without being a stable model
 * conversation identifier. Keep this registry focused on protocol boundaries;
 * session resolvers remain client-specific.
 */

/**
 * Headers whose semantics belong to a Codex Responses request and must not
 * leak into a reconstructed Chat Completions or Anthropic Messages request.
 * Capability-only values are included here even when they are not sensitive.
 */
export const codexResponsesScopedHeaderNames = [
  'openai-beta',
  'originator',
  'session-id',
  'thread-id',
  'version',
  'x-client-request-id',
  'x-oai-attestation',
  'x-openai-internal-codex-responses-lite',
  'x-openai-memgen-request',
  'x-openai-subagent',
  'x-responsesapi-include-timing-metrics'
] as const

const codexResponsesScopedHeaderNameSet = new Set<string>(codexResponsesScopedHeaderNames)
const codexResponsesScopedHeaderPrefixes = ['x-codex-'] as const

export function isCodexResponsesScopedHeaderName(name: string): boolean {
  const normalized = normalizeHeaderName(name)
  return codexResponsesScopedHeaderNameSet.has(normalized)
    || codexResponsesScopedHeaderPrefixes.some((prefix) => normalized.startsWith(prefix))
}

export function stripCodexResponsesScopedHeaders(headers: Headers): void {
  stripHeaders(headers, isCodexResponsesScopedHeaderName)
}

export const anthropicMessagesScopedHeaderNames = [
  'anthropic-beta',
  'anthropic-dangerous-direct-browser-access',
  'anthropic-version',
  'x-api-key',
  'x-app',
  'x-claude-code-agent-id',
  'x-claude-code-session-id'
] as const

const anthropicMessagesScopedHeaderNameSet = new Set<string>(anthropicMessagesScopedHeaderNames)
const anthropicMessagesScopedHeaderPrefixes = ['anthropic-', 'x-claude-code-', 'x-stainless-'] as const

export function isAnthropicMessagesScopedHeaderName(name: string): boolean {
  const normalized = normalizeHeaderName(name)
  return anthropicMessagesScopedHeaderNameSet.has(normalized)
    || anthropicMessagesScopedHeaderPrefixes.some((prefix) => normalized.startsWith(prefix))
}

export function stripAnthropicMessagesScopedHeaders(headers: Headers): void {
  stripHeaders(headers, isAnthropicMessagesScopedHeaderName)
}

export const geminiGenerateContentScopedHeaderNames = [
  'x-gemini-api-privileged-user-id',
  'x-goog-api-client',
  'x-goog-api-key',
  'x-vertex-ai-llm-request-type',
  'x-vertex-ai-llm-shared-request-type'
] as const

export type OfficialOAuthClientHeaderProfile =
  | 'openai_codex'
  | 'anthropic_claude_code'
  | 'gemini_cli'
  | 'xai_grok'

type IncomingHeaderMap = Record<string, string | string[] | undefined>

const commonOfficialOAuthHeaderNames = new Set([
  'accept',
  'accept-language',
  'content-type',
  'idempotency-key',
  'user-agent'
])

const openAIOAuthCodexHeaderNames = new Set<string>([
  ...codexResponsesScopedHeaderNames,
  'x-codex-turn-state'
])

const anthropicOAuthClaudeCodeHeaderNames = new Set<string>([
  'anthropic-beta',
  'anthropic-dangerous-direct-browser-access',
  'anthropic-version',
  'x-app',
  'x-claude-code-agent-id',
  'x-claude-code-session-id'
])

const geminiOAuthCliHeaderNames = new Set<string>([
  'api-revision',
  'x-gemini-api-privileged-user-id',
  'x-goog-api-client',
  'x-vertex-ai-llm-request-type',
  'x-vertex-ai-llm-shared-request-type'
])

const xaiOAuthGrokHeaderNames = new Set<string>([
  'x-grok-client-version',
  'x-xai-token-auth'
])

/**
 * Official subscription/OAuth adapters use a positive client-header policy.
 * API-key adapters intentionally keep using the generic safe passthrough policy.
 */
export function copyOfficialOAuthClientRequestHeaders(
  inputHeaders: IncomingHeaderMap,
  profile: OfficialOAuthClientHeaderProfile
): Headers {
  const output = new Headers()
  for (const [name, value] of Object.entries(inputHeaders)) {
    if (value === undefined || !isAllowedOfficialOAuthClientHeader(name, profile)) continue
    output.set(name, Array.isArray(value) ? value.join(', ') : value)
  }
  return output
}

function isAllowedOfficialOAuthClientHeader(
  name: string,
  profile: OfficialOAuthClientHeaderProfile
): boolean {
  const normalized = normalizeHeaderName(name)
  if (commonOfficialOAuthHeaderNames.has(normalized)) return true
  switch (profile) {
    case 'openai_codex':
      return openAIOAuthCodexHeaderNames.has(normalized) || normalized.startsWith('x-codex-')
    case 'anthropic_claude_code':
      return anthropicOAuthClaudeCodeHeaderNames.has(normalized)
        || normalized.startsWith('x-claude-code-')
        || normalized.startsWith('x-stainless-')
    case 'gemini_cli':
      return geminiOAuthCliHeaderNames.has(normalized)
    case 'xai_grok':
      return xaiOAuthGrokHeaderNames.has(normalized)
  }
}

const geminiGenerateContentScopedHeaderNameSet = new Set<string>(geminiGenerateContentScopedHeaderNames)
const geminiGenerateContentScopedHeaderPrefixes = ['x-gemini-', 'x-goog-', 'x-vertex-ai-'] as const

export function isGeminiGenerateContentScopedHeaderName(name: string): boolean {
  const normalized = normalizeHeaderName(name)
  return geminiGenerateContentScopedHeaderNameSet.has(normalized)
    || geminiGenerateContentScopedHeaderPrefixes.some((prefix) => normalized.startsWith(prefix))
}

export function stripGeminiGenerateContentScopedHeaders(headers: Headers): void {
  stripHeaders(headers, isGeminiGenerateContentScopedHeaderName)
}

function stripHeaders(headers: Headers, shouldStrip: (name: string) => boolean): void {
  for (const name of [...headers.keys()]) {
    if (shouldStrip(name)) headers.delete(name)
  }
}

function normalizeHeaderName(name: string): string {
  return name.trim().toLowerCase()
}

File: backend-go/projects/gateway/internal/gatewaydispatch/headerpolicy.go
package gatewaydispatch

import (
	"net/http"
	"strings"
)

// Header policy, migrated from upstream/header-policy.ts. A header can be
// request-dynamic or sensitive without being a stable model conversation
// identifier; session resolvers remain client-specific.

// CodexResponsesScopedHeaderNames mirrors codexResponsesScopedHeaderNames.
var CodexResponsesScopedHeaderNames = []string{
	"openai-beta",
	"originator",
	"session-id",
	"thread-id",
	"version",
	"x-client-request-id",
	"x-oai-attestation",
	"x-openai-internal-codex-responses-lite",
	"x-openai-memgen-request",
	"x-openai-subagent",
	"x-responsesapi-include-timing-metrics",
}

var codexResponsesScopedHeaderNameSet = func() map[string]struct{} {
	set := make(map[string]struct{}, len(CodexResponsesScopedHeaderNames))
	for _, name := range CodexResponsesScopedHeaderNames {
		set[name] = struct{}{}
	}
	return set
}()

var codexResponsesScopedHeaderPrefixes = []string{"x-codex-"}

// IsCodexResponsesScopedHeaderName mirrors isCodexResponsesScopedHeaderName.
func IsCodexResponsesScopedHeaderName(name string) bool {
	normalized := normalizeHeaderName(name)
	if _, ok := codexResponsesScopedHeaderNameSet[normalized]; ok {
		return true
	}
	for _, prefix := range codexResponsesScopedHeaderPrefixes {
		if strings.HasPrefix(normalized, prefix) {
			return true
		}
	}
	return false
}

// StripCodexResponsesScopedHeaders mirrors stripCodexResponsesScopedHeaders.
func StripCodexResponsesScopedHeaders(headers http.Header) {
	stripHeaders(headers, IsCodexResponsesScopedHeaderName)
}

// AnthropicMessagesScopedHeaderNames mirrors anthropicMessagesScopedHeaderNames.
var AnthropicMessagesScopedHeaderNames = []string{
	"anthropic-beta",
	"anthropic-dangerous-direct-browser-access",
	"anthropic-version",
	"x-api-key",
	"x-app",
	"x-claude-code-agent-id",
	"x-claude-code-session-id",
}

var anthropicMessagesScopedHeaderNameSet = func() map[string]struct{} {
	set := make(map[string]struct{}, len(AnthropicMessagesScopedHeaderNames))
	for _, name := range AnthropicMessagesScopedHeaderNames {
		set[name] = struct{}{}
	}
	return set
}()

var anthropicMessagesScopedHeaderPrefixes = []string{"anthropic-", "x-claude-code-", "x-stainless-"}

// IsAnthropicMessagesScopedHeaderName mirrors
// isAnthropicMessagesScopedHeaderName.
func IsAnthropicMessagesScopedHeaderName(name string) bool {
	normalized := normalizeHeaderName(name)
	if _, ok := anthropicMessagesScopedHeaderNameSet[normalized]; ok {
		return true
	}
	for _, prefix := range anthropicMessagesScopedHeaderPrefixes {
		if strings.HasPrefix(normalized, prefix) {
			return true
		}
	}
	return false
}

// StripAnthropicMessagesScopedHeaders mirrors
// stripAnthropicMessagesScopedHeaders.
func StripAnthropicMessagesScopedHeaders(headers http.Header) {
	stripHeaders(headers, IsAnthropicMessagesScopedHeaderName)
}

// GeminiGenerateContentScopedHeaderNames mirrors the Node list.
var GeminiGenerateContentScopedHeaderNames = []string{
	"x-gemini-api-privileged-user-id",
	"x-goog-api-client",
	"x-goog-api-key",
	"x-vertex-ai-llm-request-type",
	"x-vertex-ai-llm-shared-request-type",
}

// OfficialOAuthClientHeaderProfile mirrors the profile union.
type OfficialOAuthClientHeaderProfile string

const (
	OAuthHeaderProfileOpenAICodex       OfficialOAuthClientHeaderProfile = "openai_codex"
	OAuthHeaderProfileAnthropicClaude   OfficialOAuthClientHeaderProfile = "anthropic_claude_code"
	OAuthHeaderProfileGeminiCLI         OfficialOAuthClientHeaderProfile = "gemini_cli"
	OAuthHeaderProfileXAIGrok           OfficialOAuthClientHeaderProfile = "xai_grok"
)

// IncomingHeaders mirrors the Node IncomingHeaderMap.
type IncomingHeaders = http.Header

var commonOfficialOAuthHeaderNames = map[string]struct{}{
	"accept":          {},
	"accept-language": {},
	"content-type":    {},
	"idempotency-key": {},
	"user-agent":      {},
}

var openAIOAuthCodexHeaderNames = func() map[string]struct{} {
	set := make(map[string]struct{}, len(CodexResponsesScopedHeaderNames)+1)
	for _, name := range CodexResponsesScopedHeaderNames {
		set[name] = struct{}{}
	}
	set["x-codex-turn-state"] = struct{}{}
	return set
}()

var anthropicOAuthClaudeCodeHeaderNames = map[string]struct{}{
	"anthropic-beta":                          {},
	"anthropic-dangerous-direct-browser-access": {},
	"anthropic-version":                       {},
	"x-app":                                   {},
	"x-claude-code-agent-id":                  {},
	"x-claude-code-session-id":                {},
}

var geminiOAuthCliHeaderNames = map[string]struct{}{
	"api-revision":                     {},
	"x-gemini-api-privileged-user-id":  {},
	"x-goog-api-client":                {},
	"x-vertex-ai-llm-request-type":     {},
	"x-vertex-ai-llm-shared-request-type": {},
}

var xaiOAuthGrokHeaderNames = map[string]struct{}{
	"x-grok-client-version": {},
	"x-xai-token-auth":      {},
}

// CopyOfficialOAuthClientRequestHeaders mirrors
// copyOfficialOAuthClientRequestHeaders: official subscription/OAuth adapters
// use a positive client-header policy; API-key adapters keep the generic safe
// passthrough policy.
func CopyOfficialOAuthClientRequestHeaders(inputHeaders IncomingHeaders, profile OfficialOAuthClientHeaderProfile) http.Header {
	output := http.Header{}
	if inputHeaders == nil {
		return output
	}
	for name, values := range inputHeaders {
		if len(values) == 0 || !isAllowedOfficialOAuthClientHeader(name, profile) {
			continue
		}
		output.Set(name, strings.Join(values, ", "))
	}
	return output
}

func isAllowedOfficialOAuthClientHeader(name string, profile OfficialOAuthClientHeaderProfile) bool {
	normalized := normalizeHeaderName(name)
	if _, ok := commonOfficialOAuthHeaderNames[normalized]; ok {
		return true
	}
	switch profile {
	case OAuthHeaderProfileOpenAICodex:
		if _, ok := openAIOAuthCodexHeaderNames[normalized]; ok {
			return true
		}
		return strings.HasPrefix(normalized, "x-codex-")
	case OAuthHeaderProfileAnthropicClaude:
		if _, ok := anthropicOAuthClaudeCodeHeaderNames[normalized]; ok {
			return true
		}
		return strings.HasPrefix(normalized, "x-claude-code-") || strings.HasPrefix(normalized, "x-stainless-")
	case OAuthHeaderProfileGeminiCLI:
		_, ok := geminiOAuthCliHeaderNames[normalized]
		return ok
	case OAuthHeaderProfileXAIGrok:
		_, ok := xaiOAuthGrokHeaderNames[normalized]
		return ok
	}
	return false
}

var geminiGenerateContentScopedHeaderNameSet = func() map[string]struct{} {
	set := make(map[string]struct{}, len(GeminiGenerateContentScopedHeaderNames))
	for _, name := range GeminiGenerateContentScopedHeaderNames {
		set[name] = struct{}{}
	}
	return set
}()

var geminiGenerateContentScopedHeaderPrefixes = []string{"x-gemini-", "x-goog-", "x-vertex-ai-"}

// IsGeminiGenerateContentScopedHeaderName mirrors the Node predicate.
func IsGeminiGenerateContentScopedHeaderName(name string) bool {
	normalized := normalizeHeaderName(name)
	if _, ok := geminiGenerateContentScopedHeaderNameSet[normalized]; ok {
		return true
	}
	for _, prefix := range geminiGenerateContentScopedHeaderPrefixes {
		if strings.HasPrefix(normalized, prefix) {
			return true
		}
	}
	return false
}

// StripGeminiGenerateContentScopedHeaders mirrors the Node helper.
func StripGeminiGenerateContentScopedHeaders(headers http.Header) {
	stripHeaders(headers, IsGeminiGenerateContentScopedHeaderName)
}

func stripHeaders(headers http.Header, shouldStrip func(string) bool) {
	for name := range headers {
		if shouldStrip(name) {
			headers.Del(name)
		}
	}
}

func normalizeHeaderName(name string) string {
	return strings.ToLower(strings.TrimSpace(name))
}

// HeadersToObject mirrors upstream/headers.ts headersToObject.
func HeadersToObject(headers http.Header) map[string]string {
	output := make(map[string]string, len(headers))
	for name, values := range headers {
		if len(values) == 0 {
			continue
		}
		output[name] = strings.Join(values, ", ")
	}
	return output
}
